File: python/features/steps/compatibility_steps.py
"""Behave steps for compatibility mode."""

from __future__ import annotations

import json
from pathlib import Path

from behave import given, then, when

from features.steps.shared import run_cli
from taskulus.beads_write import set_test_beads_slug_sequence


@when('I run "tsk --beads list"')
def when_run_list_beads(context: object) -> None:
    run_cli(context, "tsk --beads list")


@when('I run "tsk --beads list --no-local"')
def when_run_list_beads_no_local(context: object) -> None:
    run_cli(context, "tsk --beads list --no-local")


@when('I run "tsk --beads ready"')
def when_run_ready_beads(context: object) -> None:
    run_cli(context, "tsk --beads ready")


@when('I run "tsk --beads ready --no-local"')
def when_run_ready_beads_no_local(context: object) -> None:
    run_cli(context, "tsk --beads ready --no-local")


@when('I run "tsk --beads show {identifier}"')
def when_run_show_beads(context: object, identifier: str) -> None:
    run_cli(context, f"tsk --beads show {identifier}")


@when('I run "tsk --beads create New beads child --parent bdx-epic"')
def when_run_create_beads_child(context: object) -> None:
    run_cli(context, "tsk --beads create New beads child --parent bdx-epic")


@then('beads issues.jsonl should contain "{identifier}"')
def then_beads_jsonl_contains(context: object, identifier: str) -> None:
    issues_path = context.working_directory / ".beads" / "issues.jsonl"
    contents = issues_path.read_text(encoding="utf-8")
    assert identifier in contents


@when('I run "tsk --beads create Local beads issue --local"')
def when_run_create_beads_local(context: object) -> None:
    run_cli(context, "tsk --beads create Local beads issue --local")


@when('I run "tsk --beads create Missing beads issue"')
def when_run_create_beads_missing(context: object) -> None:
    run_cli(context, "tsk --beads create Missing beads issue")


@when('I run "tsk --beads create Missing issues file"')
def when_run_create_beads_missing_issues(context: object) -> None:
    run_cli(context, "tsk --beads create Missing issues file")


@when('I run "tsk --beads create Empty beads file"')
def when_run_create_beads_empty(context: object) -> None:
    run_cli(context, "tsk --beads create Empty beads file")


@when('I run "tsk --beads create Orphan beads issue --parent bdx-missing"')
def when_run_create_beads_orphan(context: object) -> None:
    run_cli(context, "tsk --beads create Orphan beads issue --parent bdx-missing")


@when('I run "tsk --beads create Assigned beads issue --assignee dev@example.com"')
def when_run_create_beads_assigned(context: object) -> None:
    run_cli(
        context,
        "tsk --beads create Assigned beads issue --assignee dev@example.com",
    )


@when('I run "tsk --beads create Described beads issue --description Details"')
def when_run_create_beads_description(context: object) -> None:
    run_cli(context, "tsk --beads create Described beads issue --description Details")


@when('I run "tsk --beads create Beads with blanks"')
def when_run_create_beads_blank(context: object) -> None:
    run_cli(context, "tsk --beads create Beads with blanks")


@when('I run "tsk --beads create Invalid prefix"')
def when_run_create_beads_invalid_prefix(context: object) -> None:
    run_cli(context, "tsk --beads create Invalid prefix")


@when('I run "tsk --beads create Colliding beads issue"')
def when_run_create_beads_collision(context: object) -> None:
    run_cli(context, "tsk --beads create Colliding beads issue")


@when('I run "tsk --beads create Next child --parent bdx-epic"')
def when_run_create_beads_next_child(context: object) -> None:
    run_cli(context, "tsk --beads create Next child --parent bdx-epic")


@when('I run "tsk --beads update bdx-missing --status closed"')
def when_run_update_beads_missing(context: object) -> None:
    run_cli(context, "tsk --beads update bdx-missing --status closed")


@when('I run "tsk --beads update bdx-epic --status closed"')
def when_run_update_beads_success(context: object) -> None:
    run_cli(context, "tsk --beads update bdx-epic --status closed")


@when('I run "tsk --beads delete bdx-missing"')
def when_run_delete_beads_missing(context: object) -> None:
    run_cli(context, "tsk --beads delete bdx-missing")


@when('I run "tsk --beads delete bdx-task"')
def when_run_delete_beads_success(context: object) -> None:
    run_cli(context, "tsk --beads delete bdx-task")


@then('beads issues.jsonl should include assignee "{assignee}"')
def then_beads_jsonl_contains_assignee(context: object, assignee: str) -> None:
    records = _load_beads_records(_issues_path(context))
    assert any(record.get("assignee") == assignee for record in records)


@then('beads issues.jsonl should include description "{description}"')
def then_beads_jsonl_contains_description(context: object, description: str) -> None:
    records = _load_beads_records(_issues_path(context))
    assert any(record.get("description") == description for record in records)


@then('beads issues.jsonl should include status "{status}" for "{identifier}"')
def then_beads_jsonl_contains_status(
    context: object, status: str, identifier: str
) -> None:
    records = _load_beads_records(_issues_path(context))
    matches = [record for record in records if record.get("id") == identifier]
    assert matches
    assert matches[0].get("status") == status


@then('beads issues.jsonl should not contain "{identifier}"')
def then_beads_jsonl_not_contains(context: object, identifier: str) -> None:
    issues_path = _issues_path(context)
    contents = issues_path.read_text(encoding="utf-8")
    assert identifier not in contents


@given('the beads slug generator always returns "{slug}"')
def given_beads_slug_generator_returns(context: object, slug: str) -> None:
    set_test_beads_slug_sequence([slug] * 11)


@given('a beads issue with id "{identifier}" exists')
def given_beads_issue_exists(context: object, identifier: str) -> None:
    issues_path = _issues_path(context)
    record = {
        "id": identifier,
        "title": "Title",
        "issue_type": "task",
        "status": "open",
        "priority": 2,
        "created_at": "2026-02-11T00:00:00Z",
        "updated_at": "2026-02-11T00:00:00Z",
        "dependencies": [],
        "comments": [],
    }
    with issues_path.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(record) + "\n")


def _issues_path(context: object) -> Path:
    return Path(context.working_directory) / ".beads" / "issues.jsonl"


def _load_beads_records(path: Path) -> list[dict]:
    return [
        json.loads(line)
        for line in path.read_text(encoding="utf-8").splitlines()
        if line.strip()
    ]

File: rust/features/steps/compatibility_steps.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use cucumber::{given, then};
use serde_json::Value;

use kanbus::beads_write::set_test_beads_slug_sequence;
use kanbus::config::default_project_configuration;

use crate::step_definitions::initialization_steps::KanbusWorld;

fn fixture_beads_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("specs")
        .join("fixtures")
        .join("beads_repo")
        .join(".beads")
}

fn copy_dir(source: &std::path::Path, destination: &std::path::Path) {
    fs::create_dir_all(destination).expect("create beads dir");
    for entry in fs::read_dir(source).expect("read beads dir") {
        let entry = entry.expect("beads dir entry");
        let path = entry.path();
        let dest = destination.join(entry.file_name());
        if path.is_dir() {
            copy_dir(&path, &dest);
        } else {
            fs::copy(&path, &dest).expect("copy beads file");
        }
    }
}

#[given("a Kanbus project with beads compatibility enabled")]
fn given_project_with_beads_compatibility(world: &mut KanbusWorld) {
    let temp_dir = tempfile::TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("repo");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    let target_beads = repo_path.join(".beads");
    fs::create_dir_all(&target_beads).expect("beads dir");
    copy_dir(&fixture_beads_dir(), &target_beads);
    let mut configuration = default_project_configuration();
    configuration.beads_compatibility = true;
    let project_dir = repo_path.join(&configuration.project_directory);
    let issues_dir = project_dir.join("issues");
    fs::create_dir_all(&issues_dir).expect("create issues dir");
    let contents = serde_yaml::to_string(&configuration).expect("serialize config");
    fs::write(repo_path.join(".kanbus.yml"), contents).expect("write config");

    world.working_directory = Some(repo_path);
    world.temp_dir = Some(temp_dir);
}

#[given("a project directory exists")]
fn given_project_directory_exists(world: &mut KanbusWorld) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let project_dir = root.join("project");
    fs::create_dir_all(project_dir.join("issues")).expect("create issues dir");
    fs::create_dir_all(project_dir.join("events")).expect("create events dir");
}

#[then(expr = "beads issues.jsonl should contain {string}")]
fn then_beads_jsonl_contains(world: &mut KanbusWorld, identifier: String) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let path = root.join(".beads").join("issues.jsonl");
    let contents = std::fs::read_to_string(path).expect("read issues.jsonl");
    assert!(contents.contains(&identifier));
}

#[then(expr = "beads issues.jsonl should include assignee {string}")]
fn then_beads_jsonl_includes_assignee(world: &mut KanbusWorld, assignee: String) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let path = root.join(".beads").join("issues.jsonl");
    let contents = fs::read_to_string(path).expect("read issues.jsonl");
    let mut found = false;
    for line in contents.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let record: Value = serde_json::from_str(line).expect("parse record");
        if record.get("assignee").and_then(|value| value.as_str()) == Some(assignee.as_str()) {
            found = true;
            break;
        }
    }
    assert!(found);
}

#[then(expr = "beads issues.jsonl should include description {string}")]
fn then_beads_jsonl_includes_description(world: &mut KanbusWorld, description: String) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let path = root.join(".beads").join("issues.jsonl");
    let contents = fs::read_to_string(path).expect("read issues.jsonl");
    let mut found = false;
    for line in contents.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let record: Value = serde_json::from_str(line).expect("parse record");
        if record.get("description").and_then(|value| value.as_str()) == Some(description.as_str())
        {
            found = true;
            break;
        }
    }
    assert!(found);
}

#[then(expr = "beads issues.jsonl should include status {string} for {string}")]
fn then_beads_jsonl_includes_status_for(
    world: &mut KanbusWorld,
    status: String,
    identifier: String,
) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let path = root.join(".beads").join("issues.jsonl");
    let contents = fs::read_to_string(path).expect("read issues.jsonl");
    let mut found = false;
    for line in contents.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let record: Value = serde_json::from_str(line).expect("parse record");
        if record.get("id").and_then(|value| value.as_str()) == Some(identifier.as_str()) {
            let value = record.get("status").and_then(|value| value.as_str());
            assert_eq!(value, Some(status.as_str()));
            found = true;
            break;
        }
    }
    assert!(found);
}

#[then(expr = "beads issues.jsonl should not contain {string}")]
fn then_beads_jsonl_not_contains(world: &mut KanbusWorld, identifier: String) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let path = root.join(".beads").join("issues.jsonl");
    let contents = fs::read_to_string(path).expect("read issues.jsonl");
    let needle = format!("\"id\":\"{identifier}\"");
    assert!(
        !contents.contains(&needle),
        "issue id unexpectedly present in beads issues.jsonl"
    );
}

#[then(expr = "stdout should list issue {string}")]
fn then_stdout_lists_issue(world: &mut KanbusWorld, identifier: String) {
    use kanbus::ids::format_issue_key;
    let stdout = world.stdout.as_ref().expect("stdout");
    // Format the identifier for beads mode (strips project key)
    let formatted = format_issue_key(&identifier, true);
    // Check for identifier as a word boundary to avoid matching within issue type or title
    // Look for patterns like: "identifier " or " | identifier | " (porcelain)
    let patterns = vec![
        format!("{} ", identifier),
        format!(" {} ", identifier),
        format!(" | {} | ", identifier),
        format!("{} ", formatted),
        format!(" {} ", formatted),
        format!(" | {} | ", formatted),
    ];
    let found = patterns.iter().any(|pattern| stdout.contains(pattern));
    assert!(
        found,
        "Issue {} (or formatted {}) not found in stdout as identifier",
        identifier, formatted
    );
}

#[then(expr = "stdout should not list issue {string}")]
fn then_stdout_not_lists_issue(world: &mut KanbusWorld, identifier: String) {
    use kanbus::ids::format_issue_key;
    let stdout = world.stdout.as_ref().expect("stdout");
    // Format the identifier for beads mode (strips project key)
    let formatted = format_issue_key(&identifier, true);
    // Check for identifier as a word boundary
    let patterns = vec![
        format!("{} ", identifier),
        format!(" {} ", identifier),
        format!(" | {} | ", identifier),
        format!("{} ", formatted),
        format!(" {} ", formatted),
        format!(" | {} | ", formatted),
    ];
    let found = patterns.iter().any(|pattern| stdout.contains(pattern));
    assert!(
        !found,
        "Issue {} (or formatted {}) unexpectedly found in stdout",
        identifier, formatted
    );
}

#[given(expr = "the beads slug generator always returns {string}")]
fn given_beads_slug_generator_returns(_world: &mut KanbusWorld, slug: String) {
    set_test_beads_slug_sequence(Some(vec![slug; 11]));
}

#[given(expr = "a beads issue with id {string} exists")]
fn given_beads_issue_exists(world: &mut KanbusWorld, identifier: String) {
    let root = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let path = root.join(".beads").join("issues.jsonl");
    let record = serde_json::json!({
        "id": identifier,
        "title": "Title",
        "issue_type": "task",
        "status": "open",
        "priority": 2,
        "created_at": "2026-02-11T00:00:00Z",
        "updated_at": "2026-02-11T00:00:00Z",
        "dependencies": [],
        "comments": [],
    });
    let mut contents = fs::read_to_string(&path).expect("read issues.jsonl");
    if !contents.ends_with('\n') && !contents.is_empty() {
        contents.push('\n');
    }
    contents.push_str(&format!("{}\n", record.to_string()));
    fs::write(&path, contents).expect("write issues.jsonl");
}
